ipa-otptoken-import: support import from CSV file

abbra opened the pull request:
Allow import from a CSV file where fields have a fixed meaning:

"serial", "key", "type", "digits", "interval"

Fixes: https://pagure.io/freeipa/issue/7673

Pair-programmed-with: Nathaniel McCallum <npmccallum@redhat.com>
Signed-off-by: Alexander Bokovoy <abokovoy@redhat.com>

rcritten commented:
I think you should include a test with one or more invalid CSV.

Did you look at the python built-in csv module for doing the parsing?

abbra commented:
Thanks, I'll look at using a built-in csv module.

abbra commented:
I've updated the code to use csv parser and added a test to make sure we test reading invalid tokens. Note that I do not test reading invalid CSV format but there are checks for the possible exceptions thrown by that.

flo-renaud commented on ipaserver/install/ipa_otptoken_import.py:
```diff
@@ -506,6 +508,92 @@ def save(self, dest):
+        with open(filename, "r") as f:
+            reader = csv.reader(f)
+            for row in reader:
+                try:
```
You may want to add a check on the number of items in each row (the code is expecting 5), otherwise it is possible to add a CSVToken that is missing some elements.

tiran commented on ipaserver/install/ipa_otptoken_import.py:
```diff
@@ -506,6 +508,95 @@ def save(self, dest):
+        return None
+
+    def __init__(self, filename):
+        self.__keypackages = []
```
Please don't use double under. Private attribute use a single under. Double under use additional name mangling magic.

tiran commented on ipaserver/install/ipa_otptoken_import.py:
```diff
@@ -506,6 +508,95 @@ def save(self, dest):
+        self.__keypackages = []
+
+        with open(filename, "r") as f:
+            reader = csv.reader(f)
```
You can use a DictReader https://docs.python.org/3/library/csv.html#csv.DictReader to make the code more readable and stable. A dict reader will map undefined fields to None, so be prepared to raise an error when you see None.

tiran commented on ipaserver/install/ipa_otptoken_import.py:
```diff
@@ -506,6 +508,95 @@ def save(self, dest):
+                    raise ValidationError(
+                        'Provided token has incorrect number of fields')
+                try:
+                    vals = dict(zip(self._FIELDS, row))
```
This isn't necessary with a dict reader.

tiran commented on ipaserver/install/ipa_otptoken_import.py:
```diff
@@ -506,6 +508,95 @@ def save(self, dest):
+    def removed(self):
+        return self.__removed
+
+    def __init__(self, id, options):
```
``__init__`` should be defined at the top of the class definition.

tiran commented on ipaserver/install/ipa_otptoken_import.py:
```diff
@@ -506,6 +508,95 @@ def save(self, dest):
+        # Always return None as we have no key names in CSV format
+        return None
+
+    def __init__(self, filename):
```
Please move ``__init__`` before ``keyname``.

tiran commented on ipaserver/install/ipa_otptoken_import.py:
```diff
@@ -506,6 +508,95 @@ def save(self, dest):
+                        'Provided token key is not properly encoded: {e}'.
+                        format(e=str(e)))
+
+    def getKeyPackages(self):
```
PEP8 -- ``get_key_packages``

tiran commented on ipaserver/install/ipa_otptoken_import.py:
```diff
@@ -506,6 +508,95 @@ def save(self, dest):
+
+    def save(self, dest):
+        with open(dest, "w") as f:
+            writer = csv.writer(f)
```
A DictWriter also simplified this code a lot.

tiran commented on ipaserver/install/ipa_otptoken_import.py:
```diff
@@ -506,6 +508,95 @@ def save(self, dest):
+
+
+class CSVDocument(object):
+    _FIELDS = ("serial", "key", "type", "digits", "interval")
```
Why did you define different field names from IPA field names? This trick should make the code faster, smaller, and easier to understand:

```
_MAP = collections.OrderedDict([
    ('ipatokenserial', lambda v, o: v.strip()),
    ...
])

with open(...) as f:
    reader = csv.DictReader(f, list(self._MAP), restval='')
    for row in reader:
        for k, v in row.items():
            row[k] = self._MAP[k](v, options)
```

abbra commented:
I still want this to be completed but had no time yet.
